kite-data/kite-data-hive/src/main/java/org/kitesdk/data/spi/hive/HiveSchemaConverter.java

```diff
@@ -0,0 +1,357 @@
+    LinkedList<String> start = Lists.newLinkedList();
+    Collection<String[]> requiredFields = requiredFields(strategy);
+
+    List<Schema.Field> fields = Lists.newArrayList();
+    for (FieldSchema column : columns) {
+      // pass null for the initial path to exclude the table name
+      TypeInfo type = parseTypeInfo(column.getType());
+      fieldNames.add(column.getName());
+      fieldTypes.add(type);
+      fields.add(convertField(start, column.getName(), type, requiredFields));
+    }
+
+    StructTypeInfo struct = new StructTypeInfo();
+    struct.setAllStructFieldNames(fieldNames);
+    struct.setAllStructFieldTypeInfos(fieldTypes);
+
+    Schema recordSchema = Schema.createRecord(table, doc(struct), null, false);
+    recordSchema.setFields(fields);
+
+    return recordSchema;
+  }
+
+  private static Schema convert(LinkedList<String> path, String name,
+                                StructTypeInfo type,
+                                Collection<String[]> required) {
+    List<String> names = type.getAllStructFieldNames();
+    List<TypeInfo> types = type.getAllStructFieldTypeInfos();
+    Preconditions.checkArgument(names.size() == types.size(),
+        "Cannot convert struct: %s names != %s types",
+        names.size(), types.size());
+
+    List<Schema.Field> fields = Lists.newArrayList();
+    for (int i = 0; i < names.size(); i += 1) {
+      path.addLast(name);
+      fields.add(convertField(path, names.get(i), types.get(i), required));
+      path.removeLast();
+    }
+
+    Schema recordSchema = Schema.createRecord(name, doc(type), null, false);
+    recordSchema.setFields(fields);
+
+    return recordSchema;
+  }
+
+  private static Schema.Field convertField(LinkedList<String> path, String name,
+                                           TypeInfo type,
+                                           Collection<String[]> required) {
+    // filter the required fields with the current name
+    Collection<String[]> matchingRequired = filterByStartsWith(required, path, name);
+
+    Schema schema = convert(path, name, type, matchingRequired);
+    boolean isOptional = (schema.getType() == Schema.Type.UNION);
+
+    if (!isOptional && matchingRequired.size() < 1) {
+      // not already an optional union and not required, make it optional.
+      // this doesn't complain if a required field is already optional because
+      // the minimum required fields are validated by DatasetDescriptor.
+      schema  = optional(schema);
+      isOptional = true;
+    }
+
+    return new Schema.Field(name, schema, doc(type),
+        isOptional ? NULL_DEFAULT : null);
+  }
+
+  @VisibleForTesting
+  static Schema convert(LinkedList<String> path, String name,
+                        TypeInfo type, Collection<String[]> required) {
+    switch (type.getCategory()) {
+      case PRIMITIVE:
+        if (type.getClass() == charClass || type.getClass() == varcharClass) {
+          // this is required because type name includes length
```

Comment:
I don't think we want to. We decided not to add a CHAR or VARCHAR logical type to Avro and Parquet because there's no way of enforcing it and Hive has to assume that the data it reads has no idea that it is a more restrictive SQL type. I'd rather let the SQL engines handle their special types when there isn't any value to support outside.